### ISSUE TITLE
Parameters blow up when sampling

### ISSUE THREAD

**Issue by RichardSK24601** [alessiospuriomancini/cosmopower#37]
Hi,
I found this very interesting project, so I dig up my Modified CLASS with some extra parameters to see whether it can constrain these parameters effectively. I've generated the training data and trained the model for TT, TE, and EE power spectrum, and tried to constrain parameters based on the example on your website. The problem is, the sampling of the parameters seems to blow up to unrealistic numbers as the chains goes on. I don't really know what is happening here. The trained model seems able to produce proper power spectrums, but the MCMC just fails. I've included the .pkl files for these models and MCMC notebook in the attachment.

Sincerely,
Richard

PS there are essentially three new parameters introduced for the modified version of class, c0,c1, and omega_f. They described the density and scaling relation for a filud.

[models.zip](https://github.com/user-attachments/files/20816021/models.zip)

**Comment by dpiras** [comment]
Hello @RichardSK24601, thank you for your interest in CosmoPower -- glad you find it interesting!

I managed to open the notebook, but I cannot see any image with any result, so let me ask you first: do you have a plot of the accuracy of your trained models, so that we can take a look at those first?

Second, the problem might be in the sampling irrespective of CosmoPower. Did you try to run your likelihood without the emulator? Does it have the same problem or not?

Also, does the "blow up" refer to the prediction from the emulator, or to the sampling parameters? I didn't look in detail, but for instance the sampler could have a hard time if you set a prior in, say, [-0.3, 0.3] (for `omega_f`) but then start from a fiducial point at 0.5, namely outside of the prior, even with a small initial perturbation. What was the goal here?

![Image](https://github.com/user-attachments/assets/dc502074-f1e9-47fe-901d-525c307e6f28)

**Comment by RichardSK24601** [comment]
Thank you for the reply! I kinda rush through the whole process, so I didn;'t prepared testing set with my model. The loss function seems to converged pretty well. Here is an example for the training of the EE spectrum:
`Starting cosmopower_NN training, 
using 10 per cent of training samples for validation. 
Performing 5 learning steps, with 
[0.01, 0.001, 0.0001, 1e-05, 1e-06] learning rates 
[1024, 1024, 1024, 1024, 1024] batch sizes 
[1, 1, 1, 1, 1] gradient accumulation steps 
[100, 100, 100, 100, 100] patience values 
[1000, 1000, 1000, 1000, 1000] max epochs 

learning rate = 0.01, batch size = 1024
 63%|██████▎   | 631/1000 [01:25<00:50,  7.37it/s, loss=0.0286]
Validation loss = 0.025959173
Model saved.
Reached max number of epochs. Validation loss = 0.025959173
Model saved.
learning rate = 0.001, batch size = 1024
 19%|█▉        | 193/1000 [00:25<01:48,  7.42it/s, loss=0.00762]
Validation loss = 0.0072078393
Model saved.
Reached max number of epochs. Validation loss = 0.0072078393
Model saved.
learning rate = 0.0001, batch size = 1024
 10%|█         | 100/1000 [00:13<02:01,  7.43it/s, loss=0.00446]
Validation loss = 0.0042419033
Model saved.
Reached max number of epochs. Validation loss = 0.0042419033
Model saved.
learning rate = 1e-05, batch size = 1024
 10%|█         | 100/1000 [00:13<02:01,  7.38it/s, loss=0.00337]
Validation loss = 0.0032175288
Model saved.
Reached max number of epochs. Validation loss = 0.0032175288
Model saved.
learning rate = 1e-06, batch size = 1024
 10%|█         | 101/1000 [00:13<02:01,  7.41it/s, loss=0.00334]
Validation loss = 0.0032865414
Model saved.
Reached max number of epochs. Validation loss = 0.0032865414
Model saved.`

I think the blow up refers to the sampling parameters. When I check the sampled parameters in the sample array, the magnitude of these parameters gradually increased to unreasonably large value. Though I didn't test the model with a testing set, but I've plot the TT power spectrums against the one generated by CLASS. They seem to agree quite well. So I also doubt that there is something wrong with the log likelihood calculations, I'll check this later.

SIncerely,
Richard

![Image](https://github.com/user-attachments/assets/019c4bab-775b-47e6-a648-42cea89f7797)

**Comment by dpiras** [comment]
Right! I agree it looks like the emulator works fine here. I'd also recommend to only consider the spectra obtained from CLASS starting from \ell=2 (if it makes sense for your specific case), to avoid the weird behaviour at \ell=0 and \ell=1.

If the problems are in the sampling, then the issue is not `CosmoPower`; parameters blowing up might hint at the fact that you are not implementing your prior correctly. Which sampler are you using? And which priors are you setting?

**Comment by RichardSK24601** [comment]
So, I'm using the class tf_planck2018_lite provided by Cosmopower, and when I print out the log likelihoods, it includes values like -1e32 and nans.  By checking the function get_loglkl(), I included a Planck parameter. The resulting log likelihood became -inf instead of nans. Probably because this class is not suitable for this modified model?

**Comment by dpiras** [comment]
Hmm, I think there are some things we should check. 

First, did you try the `get_loglkl` function on the CosmoPower emulators we provide, to check you don't get any nans or infs?

Secondly, did you train on the log of the spectra, and what do you provide as input to the `get_loglkl` function?

**Comment by RichardSK24601** [comment]
There is no nan for the emulators provided by Cosmopower. I do trained my model with log10 of the power spectrum, and I find that I can get away these nan if I changed the line 277 of tf_planck2018_lite.py
`Clte= self.te_emu_model.predictions_tf(cosmo_params)` to `Clte= self.te_emu_model.ten_to_predictions_tf(cosmo_params)`.

But still the sampling parameters goes to pretty large value, maybe there is something wrong with TE, EE prediction or the low l prediction of power spectrum. I'm trying to write a new function to calculate the log likelihood.

**Comment by dpiras** [comment]
So, do you still get nans/infs in the likelihood (even before running the MCMC, just at the likelihood level) with your new emulator?

**Comment by RichardSK24601** [comment]
There is no nans anymore in the likelihood level.

**Comment by dpiras** [comment]
Ah, that's nice! 

Which values of the likelihood do you typically get?

Which sampler do you use then to sample your posterior? And which prior do you set on the parameters?

**Comment by dpiras** [comment]
@RichardSK24601 Did you manage to solve your problems with the sampler?

**Comment by RichardSK24601** [comment]
Sorry for the late reply, I've been working on other projects. The value that I typically get is -1e+32. I'm currently still using the affine library to do the sampling. But still, I suspect it is a likelihood issue. 
I'm currently using this set of prior:
`parameters_and_priors = {'omega_b':      [0.001, 0.04, 'uniform'],
                         'omega_cdm':    [0.005, 0.99,  'uniform'],
                         'h':            [0.2,   1.0,   'uniform'],
                         'tau_reio':     [0.01,  0.8,   'uniform'],
                         'n_s':          [0.9,   1.1,   'uniform'],
                         'ln10^{10}A_s': [1.61,  3.91,  'uniform'],
                         'A_planck':     [1.2,   0.9,  'uniform'],
                          }

# instantiation
tf_planck = cp.tf_planck2018_lite(parameters=parameters_and_priors, 
                                  tf_planck2018_lite_path=tf_planck2018_lite_path,
                                  tt_emu_model=tt_emu_model,
                                  te_emu_model=te_emu_model,
                                  ee_emu_model=ee_emu_model
                                  )

# initial points
FIDUCIAL = np.reshape(np.array([ 0.022242,  0.11977,  0.673,  0.055,  0.9658,  3.0753,  1]).astype('float32'), (1, 7))
EPSILON  = np.reshape(np.array([ 1E-4,      1E-4,     1E-2,   1E-3,   1E-3,    1E-3,    1E-4]).astype('float32'), (1,7))`

**Comment by alessiospuriomancini** [comment]
Hi @RichardSK24601, thanks for your interest in CosmoPower. 

Please check your 'A_planck' prior. The original one in the [notebook](https://github.com/alessiospuriomancini/cosmopower/blob/main/notebooks/likelihoods_notebooks/tf_planck2018_lite.ipynb) is 

    'A_planck':     [1.0,   0.01,  'gaussian'],

whereas you have 

    'A_planck': [1.2, 0.9, 'uniform'],

which is not possible and gives an error.

**Comment by dpiras** [comment]
Adding to what @alessiospuriomancini said, I think likelihood values of -1e+32 sound suspicious. Could you perhaps share a spectrum and the likelihood function you use when you have time, so that we can check which values we get?

**Comment by alessiospuriomancini** [comment]
@RichardSK24601 I'm closing this issue as resolved, feel free to re-open if needed
